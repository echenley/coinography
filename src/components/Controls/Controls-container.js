// @flow

import { connect, type Connector } from 'react-redux'
import { updateChartType, updateCurrency, updateGranularity } from '../../actions'
import { getChartType, getCurrency, getGranularity } from '../../selectors'
import Controls, { type Props } from './Controls'
import type { State } from '../../types'

function mapStateToProps(state: State): * {
  return {
    chartType: getChartType(state),
    currency: getCurrency(state),
    granularity: getGranularity(state),
  }
}

const connector: Connector<{}, Props> = connect(mapStateToProps, {
  updateChartType,
  updateCurrency,
  updateGranularity,
})
export default connector(Controls)
