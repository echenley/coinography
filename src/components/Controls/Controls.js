// @flow

import React from 'react'
import { View, StyleSheet } from 'react-native'
import { Picker } from 'native-base'
import type { ChartType, Currency, Granularity } from '../../types'

export type Props = {
  chartType: ChartType,
  currency: Currency,
  granularity: Granularity,
  updateChartType: (chartType: ChartType) => mixed,
  updateCurrency: (currency: Currency) => mixed,
  updateGranularity: (granularity: Granularity) => mixed,
}

const currencies: Array<Currency> = ['BTC-USD', 'ETH-USD']

const granularities: Array<Granularity> = [60, 300, 900, 3600, 21600, 86400]
const granularityLabels = {
  '60': '1m',
  '300': '5m',
  '900': '15m',
  '3600': '1h',
  '21600': '6h',
  '86400': '1d',
}

const chartTypes = ['line', 'candle']
const chartTypeLabels = {
  line: 'Line',
  candle: 'Candle',
}

const Controls = ({
  chartType,
  currency,
  granularity,
  updateChartType,
  updateCurrency,
  updateGranularity,
}: Props) => {
  return (
    <View style={styles.controls}>
      <Picker
        iosHeader="Chart Type"
        mode="dropdown"
        onValueChange={updateChartType}
        selectedValue={chartType}
        style={styles.picker}
        supportedOrientations={['portrait', 'landscape']}
      >
        {chartTypes.map(type => (
          <Picker.Item key={type} label={chartTypeLabels[type]} value={type} />
        ))}
      </Picker>
      <Picker
        iosHeader="Granularity"
        mode="dropdown"
        onValueChange={updateGranularity}
        selectedValue={granularity}
        style={styles.picker}
        supportedOrientations={['portrait', 'landscape']}
      >
        {granularities.map(g => (
          <Picker.Item key={g} label={granularityLabels[g]} value={g} />
        ))}
      </Picker>
      <Picker
        iosHeader="Currency Pair"
        mode="dropdown"
        onValueChange={updateCurrency}
        selectedValue={currency}
        style={styles.picker}
        supportedOrientations={['portrait', 'landscape']}
      >
        {currencies.map(c => <Picker.Item key={c} label={c} value={c} />)}
      </Picker>
    </View>
  )
}

const styles = StyleSheet.create({
  picker: {
    display: 'flex',
    flex: 1,
    justifyContent: 'center',
    marginLeft: 2,
    marginRight: 2,
    backgroundColor: 'rgba(0, 0, 0, 0.05)',
  },
  controls: {
    display: 'flex',
    justifyContent: 'center',
    flexDirection: 'row',
    padding: 10,
  },
})

export default Controls
