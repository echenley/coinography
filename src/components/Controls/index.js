// @flow

import Controls from './Controls-container'
export default Controls
