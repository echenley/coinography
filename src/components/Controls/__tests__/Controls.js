import * as React from 'react'
import { shallow } from 'enzyme'
import { Picker } from 'native-base'
import Controls from '../Controls'

function renderComponent(props) {
  return shallow(
    <Controls
      chartType="candle"
      currency="BTC-USD"
      granularity={300}
      updateChartType={() => {}}
      updateCurrency={() => {}}
      updateGranularity={() => {}}
      {...props}
    />
  )
}

describe('<Controls />', () => {
  it('renders correctly', () => {
    expect(renderComponent().getElement()).toMatchSnapshot()
  })

  it('calls update functions', () => {
    const updateChartType = jest.fn()
    const updateCurrency = jest.fn()
    const updateGranularity = jest.fn()
    const component = renderComponent({
      updateChartType,
      updateCurrency,
      updateGranularity,
    })

    expect(updateChartType).not.toHaveBeenCalled()
    expect(updateGranularity).not.toHaveBeenCalled()
    expect(updateCurrency).not.toHaveBeenCalled()

    component
      .find(Picker)
      .at(0)
      .prop('onValueChange')('line')
    expect(updateChartType).toHaveBeenCalledWith('line')

    component
      .find(Picker)
      .at(1)
      .prop('onValueChange')(3600)
    expect(updateGranularity).toHaveBeenCalledWith(3600)

    component
      .find(Picker)
      .at(2)
      .prop('onValueChange')('ETH-USD')
    expect(updateCurrency).toHaveBeenCalledWith('ETH-USD')
  })
})
