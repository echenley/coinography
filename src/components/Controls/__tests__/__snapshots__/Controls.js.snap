// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Controls /> renders correctly 1`] = `
<View
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "padding": 10,
    }
  }
>
  <Styled(PickerNB)
    iosHeader="Chart Type"
    mode="dropdown"
    onValueChange={[Function]}
    selectedValue="candle"
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.05)",
        "display": "flex",
        "flex": 1,
        "justifyContent": "center",
        "marginLeft": 2,
        "marginRight": 2,
      }
    }
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    virtual={undefined}
  >
    <Item
      label="Line"
      value="line"
    />
    <Item
      label="Candle"
      value="candle"
    />
  </Styled(PickerNB)>
  <Styled(PickerNB)
    iosHeader="Granularity"
    mode="dropdown"
    onValueChange={[Function]}
    selectedValue={300}
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.05)",
        "display": "flex",
        "flex": 1,
        "justifyContent": "center",
        "marginLeft": 2,
        "marginRight": 2,
      }
    }
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    virtual={undefined}
  >
    <Item
      label="1m"
      value={60}
    />
    <Item
      label="5m"
      value={300}
    />
    <Item
      label="15m"
      value={900}
    />
    <Item
      label="1h"
      value={3600}
    />
    <Item
      label="6h"
      value={21600}
    />
    <Item
      label="1d"
      value={86400}
    />
  </Styled(PickerNB)>
  <Styled(PickerNB)
    iosHeader="Currency Pair"
    mode="dropdown"
    onValueChange={[Function]}
    selectedValue="BTC-USD"
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.05)",
        "display": "flex",
        "flex": 1,
        "justifyContent": "center",
        "marginLeft": 2,
        "marginRight": 2,
      }
    }
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    virtual={undefined}
  >
    <Item
      label="BTC-USD"
      value="BTC-USD"
    />
    <Item
      label="ETH-USD"
      value="ETH-USD"
    />
  </Styled(PickerNB)>
</View>
`;
